### ISSUE TITLE
Add an immutable HashMap to the javaslang.collection package

### ISSUE THREAD

**Issue by danieldietrich** [javaslang/javaslang#170]
Interesting read: http://javarevisited.blogspot.de/2011/02/how-hashmap-works-in-java.html

**Comment by danieldietrich** [comment]
Maps work best backed by a Trie, see "Purely Functional Data Structures", p. 163 ff.

**Comment by danieldietrich** [comment]
Good read on functional/immutable Map: http://blog.ezyang.com/2010/03/the-case-of-the-hash-array-mapped-trie/

**Comment by danieldietrich** [comment]
Ideal Hash Trees (Phil Bagwell/Clojure): http://lampwww.epfl.ch/papers/idealhashtrees.pdf

(See also wikipedia [Trie](http://en.wikipedia.org/wiki/Trie) and [Hash array mapped trie](http://en.wikipedia.org/wiki/Hash_array_mapped_trie))

**Comment by danieldietrich** [comment]
Fast mergable integer maps: http://citeseer.ist.psu.edu/viewdoc/download?doi=10.1.1.37.5452&rep=rep1&type=pdf

**Comment by danieldietrich** [comment]
@ignasi35 Here I provided some links for the Map implementation. I think, Map is one of the most complex collections to implement.

Without having read on Maps yet, a naive impl would be to use a balanced binary search tree with (informally) `Entry(key.hashCode, List(values))`, where `equals` and `compareTo` of an `Entry` just compare the `key.hashCode` and not the values. Given that, a tree node could be found just by the `hashCode` and the value could be obtained from the associated List. We need a List of values at each node because of hash collisions.

But my understanding is that `HashMap.get()` performs in O(1). The above impl would implement in `O(log n)`...

Better implementations use the so called `Trie` data structure with 'marked' edges (and values at the leafs only?). This has to be further investigated... How does Java implement (mutable) HashMap? How does Scala implement HashMap?

**Comment by ignasi35** [comment]
Some more tips on the subject. http://stackoverflow.com/a/9054660

**Comment by ruslansennov** [comment]
@danieldietrich  please review this HAMT implementation ruslansennov/javaslang@b9fcfb185175b02be7ddc99e8641309f178f6942
What other methods need to be implemented?

**Comment by danieldietrich** [comment]
@ruslansennov Wow, Ruslan, I'm really impressed. This is \*exactly\* what we need. This is the best implementation I can imagine - head to head with Clojure and Scala.

I will add some line notes in a few minutes.

Here is an overview of the design of the upcoming collections:

![javaslang-classes](https://cloud.githubusercontent.com/assets/743833/9027154/491a0d3e-394a-11e5-8568-39f6db840b8d.png)

We need to create a `Map` interface. The additional operations compared to `Traversable` to start with are:

```java
interface Map<K, V> extends Traverable<Entry<K, V>>

    boolean containsKey(K key);

    boolean containsValue(V value);

    // we will add this when we have Set
    // Set<Entry<K, V>> entrySet();

    V get(T key);

    V getOrDefault(Object key, V defaultValue);

    // we will add this when we have Set
    // Set<K> keySet();

    Map<K, V> put(K key, V value);

    Map<K, V> remove(K key);

    Traversable<V> values();

    final class Entry<K, V> implements Serializable {

        private static final long serialVersionUID = 1L;

        public final K key;
        public final V value;

        public Entry(K key, V value) {
            this.key = key;
            this.value = value;
        }

        @Override
        public boolean equals() {
            if (o == this) {
                return true;
            } else if (o instanceof Entry) {
                final Entry<?, ?> that = (Entry<?, ?>) o;
                return Objects.equals(this.key, that.key) && Objects.equals(this.value, that.value);
            }
        }

        @Override
        public int hashCode() {
            return Objects.hash(key) * 31 + Objects.hash(value);
        }

        @Override
        public String toString() {
            return "(" + key + ", "+ value + ");
        }
    }
}
```

In the image attached above, I put separate `Trie`. Does this make sense? Can there be a general purpose Trie? In my understanding, the HAMT Trie implementation is very specific to the HashMap application.

My plan was to not bother the user with implementation details. That means, I would just provide a `Map` and a `HashMap`. The `HashMap` internally is a HAMT but the user does not see this. It is just works as sophisticated as possible (like HAMT backed HashMap).

**Implementing all the `Traversable` is lot of work. We may in a first step provide a `Map` interface, which does not extends `Traversable`, so we can pull a version into the master branch.**

**Comment by ruslansennov** [comment]
I also think that we should not implement all methods of Traversable.

**Comment by danieldietrich** [comment]
Yes you are right. `Traversable` works good for (linear) data structures like sequences and sets. For more complex (nested) structures, like Trees and Maps, it is confusing, overloaded and seems to be a little bit faked.

Conversion between `Traversable` and `Map` can be done by calling `Map.entrySet()` and `HashMap.of(Iterable<Entry>)`. Also we will create `<K,V> TranversableOnce.toHashMap(Function<? super T, ? extends Entry<? extends K, ? extends V>)`.

Map should implement `TraversableOnce` and the upcoming `FilterMonadic` (map, flatMap, filter, flatten). This should be enough for now. Additionally we should implement the features of Java's [Map](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html).

Btw. I like to use `size()` here instead of Traversable's `length()`.

**Comment by danieldietrich** [comment]
@ruslansennov If it is ok for you, I will assign you to this task.

**Comment by danieldietrich** [comment]
@ruslansennov Scala's `Map` does implement `Traversable<Entry>`, i.e. take, drop, reduce, sum, ...
I target this also for Javaslang. Maby we can implement working _placeholders_ for most operations:

```java
HashMap<...> someOperation() {
    return toList().someOperation().toHashMap(Function.identity);
}
```

**Comment by ruslansennov** [comment]
OK, see #410

**Comment by danieldietrich** [comment]
Our implementation of the HAMT (which @ruslansennov contributed) is based on:

* https://github.com/mattbierner/hamt/blob/master/lib/hamt.kep
* https://github.com/exclipy/pdata/blob/master/Data/HamtMap.hs

**Comment by danieldietrich** [comment]
## State of the Traversables

I'm currently getting the Map vs Set vs Seq interfaces right. With my refactoring it was clear that some methods have to move from the Traversable interface to Seq because the signatures in Map differ.

Example:

```java
// Seq
<U> Seq<Tuple2<T, U>> zip(Iterable<U> that);
<U> Seq<Tuple2<T, U>> zipAll(Iterable<U> that, T thisElem, U thatElem);
Seq<Tuple2<T, Integer>> zipWithIndex();

// Map
<U> Map<Tuple2<K, V>, U> zip(Iterable<U> that);
<U> Map<Tuple2<K, V>, U> zipAll(Iterable<U> that, Entry<K, V> thisElem, U thatElem);
Map<Tuple2<K, V>, Integer> zipWithIndex();
```

Also we can only `reverse()` something, which has an order. Reversing a Set makes no sense.

Currently I revise the Traversable interface to get Seq, Set and Map method signatures right.

My strategy is, to follow the types. If in doubt, I have a look at Haskell, Scala et al.

**Comment by danieldietrich** [comment]
Also we have the opportunity to get rid of null.

**Example:** `Map.get(key)` should return `Option<V>` instead of `V`. `getOption(key)` will be removed accordingly. This is also how Scala operates.
```

**Comment by danieldietrich** [comment]
`Seq.get(int)` may by ambiguous with `Map.get(key)`, if `key` is of type `Integer`. Therefore we remove `Seq.get(int)`. Because `Seq<T>` is an `IntFunction<T>` we use `apply(int)` instead.

**Comment by ruslansennov** [comment]
This looks like a revolution. Please review my PR before :)

**Comment by danieldietrich** [comment]
will do now :-)

**Comment by danieldietrich** [comment]
Fixed with #609
